packages/bsky/src/notifications.ts

```diff
@@ -0,0 +1,285 @@
+import axios from 'axios'
+import Database from './db'
+import { Notification } from './db/tables/notification'
+import { AppBskyEmbedImages, AtUri } from '@atproto/api'
+import { Insertable } from 'kysely'
+
+const PUSH_NOTIF_SERVER_BASE_URL = 'https://push.bsky.app'
+const PUSH_NOTIF_SERVER_ENDPOINT = '/api/push'
+export const PUSH_NOTIF_SERVER_URL =
+  PUSH_NOTIF_SERVER_BASE_URL + PUSH_NOTIF_SERVER_ENDPOINT
+export const BSKY_APP_ID = 'xyz.blueskyweb.app'
+type Platform = 'ios' | 'android' | 'web'
+type PushNotification = {
+  tokens: string[]
+  platform: 1 | 2 // 1 = ios, 2 = android
+  title: string
+  message: string
+  topic: string
+  data?: {
+    [key: string]: string
+  }
+}
+type InsertableNotif = Insertable<Notification>
+
+export class NotificationServer {
+  constructor(public db: Database) {}
+  static creator() {
+    return (db: Database) => new NotificationServer(db)
+  }
+
+  async getUserTokens(did: string) {
+    const userTokens = await this.db.db
+      .selectFrom('notification_push_token')
+      .where('did', '=', did)
+      .selectAll()
+      .execute()
+
+    return userTokens
+  }
+
+  async prepareNotifsToSend(notifications: InsertableNotif[]) {
+    const notifsToSend: PushNotification[] = []
+
+    for (const notif of notifications) {
+      const { did: userDid } = notif
+      const userTokens = await this.getUserTokens(userDid)
+      // if user has no tokens, skip
+      if (!userTokens || userTokens.length === 0) {
+        continue
+      }
+
+      for (const t of userTokens) {
+        const { appId, platform, token } = t
+        // get title and message of notification
+        const attr = await this.getNotificationDisplayAttributes(notif)
+        // if no title or body, skip
+        if (!attr) {
+          continue
+        }
```

Comment:
Looks like `attr` will always be defined, unless `getNotificationDisplayAttributes` throws, in which case this loop will exit, right?

Do we want to catch and filter out invalid notifications so we can send valid ones, or just exit the whole notification process?

Reply:
There were times I was getting no `attr` like it was an empty object `{}`. Not 100% sure why